wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -357,705 +401,903 @@ The different scenarios and the expected outcomes are,
+Exceptions are unexpected events or errors that occur during the execution of a program, disrupting the normal flow of instructions. These are typically conditions that a program cannot anticipate or recover from easily, such as a division by zero, accessing an invalid index in an array, or trying to open a file that doesn’t exist. When an exception occurs, the program may terminate abruptly unless the exception is properly handled.
 
-Examples:
-- `NullPointerException` (accessing a null reference)
-- `IllegalArgumentException` (invalid function argument)
-- `IOException` (I/O operation failure)
+```kotlin
+fun divideNumbers(numerator: Int, denominator: Int): Int {
+  if (denominator == 0) throw IllegalArgumentException("Denominator cannot be zero")
+  return numerator / denominator
+}
+```
 
-This allows to catch and handle issues, allowing the program to continue running or fail in a controlled manner.
+In this example, if the denominator is zero, an `IllegalArgumentException` is thrown. This is a standard way to handle situations where continuing the program as usual doesn’t make sense due to an error in the input.
 
-For instance, to handle division by zero error.
+**Testing Exceptions:**
 
-```
-fun divide(a: Int, b: Int): Int {
-    if (b == 0) throw ArithmeticException("Division by zero")
-    return a / b
+The primary focus when testing exceptions is ensuring that the correct exception is thrown in the appropriate circumstances and that the program can handle it gracefully.
+
+**Test:**
+
+```kotlin
+@Test
+fun testDivideNumbers_forZeroDenominator_throwsIllegalArgumentException() {
+  val exception = assertThrows<IllegalArgumentException> {
+    divideNumbers(10, 0)
+  }
+  assertThat(exception).hasMessageThat().contains("Denominator cannot be zero")
 }
 ```
 
-**Error Handling:**
+This test verifies that the divideNumbers function throws an IllegalArgumentException when the denominator is zero. It also checks that the exception message contains the expected text. Testing exceptions involves confirming that the application correctly identifies and handles these unexpected situations.
 
-Errors are more severe issues indicating fundamental problems with the program's environment or state, often beyond the program's control. In Java/Kotlin, these are represented by classes extending Error, like `OutOfMemoryError` or `StackOverflowError`.
+### Domain Errors
 
-This allows to manage situations where the application may not continue running properly, ensuring the program does not enter an unstable state.
+Domain errors are errors related to the business logic or rules of the application, rather than technical issues like those covered by exceptions. These errors occur when the data or conditions within the domain do not meet the specific rules or constraints defined by the business. Domain errors are expected conditions and are handled within the normal flow of the application, rather than through exceptions.
```

Comment:
Okay, updated it to - "Domain errors are errors related to **the logic** or rules of the application, rather than technical issues like those covered by exceptions. These errors occur when the data or conditions within the domain do not meet the specific rules or constraints **defined during application design**. Domain errors are expected conditions and are handled within the normal flow of the application, rather than through exceptions.
"

Reply:
Ah much cleaner--thanks!